crates/shared/src/utils.rs

```diff
@@ -148,37 +153,189 @@ impl<Types: NodeType, V: StaticVersionType + 'static> Stream for EventServiceStr
     ) -> Poll<Option<Self::Item>> {
         match &mut self.connection {
             Left(connection) => {
-                let Poll::Ready(next) = connection.poll_next_unpin(cx) else {
-                    return Poll::Pending;
-                };
+                let next = ready!(connection.poll_next_unpin(cx));
                 match next {
                     Some(Ok(event)) => Poll::Ready(Some(event)),
                     Some(Err(err)) => {
-                        error!("Error in the event stream: {err:?}");
+                        warn!(?err, "Error in event stream");
                         Poll::Pending
                     }
                     None => {
+                        warn!("Event stream ended, attempting reconnection");
                         let fut = Self::connect_inner(self.api_url.clone());
                         let _ = std::mem::replace(&mut self.connection, Right(Box::pin(fut)));
+                        cx.waker().wake_by_ref();
                         Poll::Pending
                     }
                 }
             }
             Right(reconnect_future) => {
-                let Poll::Ready(ready) = reconnect_future.poll_unpin(cx) else {
-                    return Poll::Pending;
-                };
-                match ready {
+                let next = ready!(reconnect_future.poll_unpin(cx));
+                match next {
+                    Err(err) => {
+                        error!(?err, "Error while reconnecting, giving up");
+                        Poll::Ready(None)
```

Comment:
Also I don't think this is necessarily the behavior we want. If we get disconnected, it's likely because the query node restarted, which takes some time. If we try again right away, we are likely to fail again, but if we keep trying, we are likely to eventually succeed.

I think since the builder becomes largely useless if this stream ends, it's not a good idea to give up here. Instead we can keep trying (and complaining loudly) until we successfully reconnect

Reply:
This is the case where the API became reachable, but responds with something unexpected.
Do you think we should try reconnecting in such case as well?

Reply:
Meh maybe it was reachable for a sec for the healthcheck to respond, but became unreachable again when we tried to establish the WebSocket connection. Who knows. Honestly I think we should retry no matter what because it's the safest failure mode, and as long as we are logging with error, someone will see it and manually intervene if it doesn't automatically recover. But indefinite retries here gives us the best change for a temporarily blip to resolve before someone gets woken up 😆

Reply:
Sure, I'll route back to retry then